### PR TITLE
language server protocol (VS Code integration)

### PR DESCRIPTION
**Pull request by eed3si9n**
This is a work in progress branch to support Language Server Protocol (Visual Studio Code integration) on sbt server.

The initial goal is to invoke `compile` on save. The details are written up in this blog post http://eed3si9n.com/scala-language-server-using-sbt

Depends on:
- [x] https://github.com/sbt/util/pull/128
- [x] https://github.com/sbt/sbt/pull/3556
- [x] https://github.com/sbt/zinc/pull/413
- [x] https://github.com/eed3si9n/sjson-new/issues/83

### CODE REVIEW

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by smarter** [comment]
Wow, nice! This might be enough motivation for me to finish upgrading dotty to sbt 1. I think it'd be very cool if JSON-RPC was the only protocol used to communicate with sbt server mode. We could define a build server protocol that complements the language server protocol.

**Comment by eed3si9n** [comment]
> I think it'd be very cool if JSON-RPC was the only protocol used to communicate with sbt server mode. We could define a build server protocol that complements the language server protocol.

I agree. While there's no clients or documentations yet, we should try to switch to JSON-RPC as the wire protocol if we can.

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by eed3si9n** [comment]
With f6bd70b, VS Code now talks to sbt server and invokes `compile` at the root level.

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

File: build.sbt [hunk 1]
```diff
@@ -387,6 +387,41 @@ lazy val sbtProj = (project in file("sbt"))
   )
   .configure(addSbtCompilerBridge)
 
+lazy val vscodePlugin = (project in file("vscode-sbt-scala"))
```
Comment (smarter) [inline build.sbt:414]: Feel free to take some inspiration from https://github.com/lampepfl/dotty/blob/657385ba36eb138f1f78413311b7c108d3dc8ac2/project/Build.scala#L891-L936

File: vscode-sbt-scala/client/package.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  },
+  "description": "VSCode part of the Scala language server",
+  "engines": {
+    "vscode": "^1.16.0"
+  },
+  "categories": [
+    "Other"
+  ],
+  "activationEvents": [
+    "onLanguage:plaintext"
```
Comment (smarter) [inline vscode-sbt-scala/client/package.json]: That doesn't seem right.`"workspaceContains:build.sbt"` would be better, but unfortunately things tend to break when two language servers are active at the same time in vscode (see https://github.com/dragos/dragos-vscode-scala/pull/47). One solution would be to add code in the vscode client extension to not start the server if either `.ensime` or `.dotty-ide.json` are present. (And then I can modify the dotty language server to start and talk to an instance of the sbt language server).

Comment (eed3si9n) [reply vscode-sbt-scala/client/package.json]: Good catch. `"onLanguage:plaintext"` is a leftover from their sample code.

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

File: main/src/main/scala/sbt/internal/server/LanguageServerProtocol.scala [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import sbt.internal.protocol.codec._
+import sbt.internal.langserver._
+import sbt.internal.util.ObjectEvent
+
+private[sbt] case class LangServerError(code: Long, message: String) extends Throwable(message)
+
+/**
+ * Implements Language Server Protocol <https://github.com/Microsoft/language-server-protocol>.
+ */
+private[sbt] trait LanguageServerProtocol extends CommandChannel {
```
Comment (eed3si9n) [inline main/src/main/scala/sbt/internal/server/LanguageServerProtocol.scala:20]: @dwijnand The key logic should mostly be in this file.

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by guillaumebort** [comment]
Interesting. I did the same thing as an experiment, but I just used the sbt server protocol directly. Works pretty well.

**Comment by kiritsuku** [comment]
I tried to run
```
sbt:sbtRoot> vscodePlugin/compile
Warning: Unknown Scalafmt version 1.2.0; using 1.0 interface
[warn] Credentials file /home/sschaef/.bintray/.credentials does not exist

> vscode-sbt-scala@0.0.1 compile /home/sschaef/dev/scala/sbt/vscode-sbt-scala
> tsc -p client/tsconfig.json && cd server && npm run installServer && cd .. && tsc -p server/tsconfig.json

client/node_modules/vscode-languageclient/lib/client.d.ts(1,749): error TS2307: Cannot find module 'vscode'.
client/node_modules/vscode-languageclient/lib/codeConverter.d.ts(1,23): error TS2307: Cannot find module 'vscode'.
client/node_modules/vscode-languageclient/lib/main.d.ts(5,28): error TS2307: Cannot find module 'vscode'.
client/node_modules/vscode-languageclient/lib/protocolCompletionItem.d.ts(1,23): error TS2307: Cannot find module 'vscode'.
client/node_modules/vscode-languageclient/lib/protocolConverter.d.ts(1,23): error TS2307: Cannot find module 'vscode'.
client/node_modules/vscode-languageclient/lib/workspaceFolders.proposed.d.ts(1,77): error TS2307: Cannot find module 'vscode'.
client/src/extension.ts(6,45): error TS2307: Cannot find module 'vscode'.

npm ERR! Linux 4.4.0-96-generic
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "run" "compile"
npm ERR! node v6.11.3
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! vscode-sbt-scala@0.0.1 compile: `tsc -p client/tsconfig.json && cd server && npm run installServer && cd .. && tsc -p serv
er/tsconfig.json`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the vscode-sbt-scala@0.0.1 compile script 'tsc -p client/tsconfig.json && cd server && npm run installServer &&
cd .. && tsc -p server/tsconfig.json'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the vscode-sbt-scala package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     tsc -p client/tsconfig.json && cd server && npm run installServer && cd .. && tsc -p server/tsconfig.json
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs vscode-sbt-scala
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls vscode-sbt-scala
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/sschaef/dev/scala/sbt/vscode-sbt-scala/npm-debug.log
```

What is the module `vscode` that can't be found?

**Comment by dwijnand** [comment]
@sschaef What part of `tsc -p client/tsconfig.json && cd server && npm run installServer && cd .. && tsc -p server/tsconfig.json` fails on your system?

**Comment by kiritsuku** [comment]
The first command already fails:
```
% tsc -p client/tsconfig.json
client/node_modules/vscode-languageclient/lib/client.d.ts(1,749): error TS2307: Cannot find module 'vscode'.
```

I did a `npm install vscode` beforehand and also a `sudo npm install -g vscode` but `tsc` still can't find the `vscode` module.

**Comment by kiritsuku** [comment]
Ok, I fixed the problem. On sbt startup there is a `npm install` run on the vscode project. I removed the `node_modules` folder and restart sbt. Now everything works perfectly fine.

**Comment by kiritsuku** [comment]
In the vscode part, there are two servers. One is in `vscode-sbt-scala/client/server` and the other is in `vscode-sbt-scala/server`. Why are there two of them, isn't one enough?

What is this `package-lock.json` file doing? Is it required?

**Comment by eed3si9n** [comment]
@sschaef The example extension was set up that way with "client" and "server" at the top-level, and the compiled server code copied into client.
They represent two different Node instances so I think it kind of make sense to me that the source directory is separate. In my implementation "server" just converts stdio to socket.

**Comment by eed3si9n** [comment]
I am not familiar with Node development, so I'm not sure, but I am guessing that `package-lock.json` is a dependency lock file that pins down the list of dependencies. If so I think it would make sense to commit it.

**Comment by kiritsuku** [comment]
Ok, makes sense now. `package-lock.json` is indeed intended to be committed: https://docs.npmjs.com/files/package-lock.json

**Review by dwijnand — APPROVED** [review]
LGTM, pending Travis CI passing.

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3524/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

File: server.md [hunk 4]
```diff
@@ -1,11 +1,34 @@
...
+### sbt/exec
+
+```json
+{ "jsonrpc": "2.0", "id": 1, "method": "sbt/exec", "params": { "commandLine": "compile" } }
+```
+
+### sbt/setting
+
+```json
+{ "jsonrpc": "2.0", "id": 1, "method": "sbt/setting", "params": { "setting": "root/name" } }
```
Comment (dwijnand) [inline server.md:19]: sbt/query?

**Comment by kiritsuku** [comment]
Is the `textDocument/didSave` notification already working for you? I don't receive any notifications on the server side yet.

**Comment by eed3si9n** [comment]
@sschaef Yea, it works for me. You have to run `reset.sh` to make sure you are using freshly baked sbt tho.

File: server.md (hunk 4 above)
Comment (eed3si9n) [reply server.md:19]: I thought about keeping the name query in there, but the Microsoft style naming for query is using just the nouns it seems. See https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#workspace_symbol

**Comment by kiritsuku** [comment]
I think I have a problem with vscode (probably not using it correctly), because sbt doesn't even receive a notification event. The initialization routine and the shutdown event is working but nothing else. Can you edit arbitrary files or does it work only for specific files?

**Comment by eed3si9n** [comment]
It works for me using `Hello.scala` not sure if it would for other extensions.

**Comment by smarter** [comment]
@sschaef: Right now it's restricted to .scala and .java files: https://github.com/eed3si9n/sbt/blob/1ad9360e3e9446eed325fd7259c41b59541393dc/vscode-sbt-scala/client/src/extension.ts and requires a build.sbt file: https://github.com/eed3si9n/sbt/blob/1ad9360e3e9446eed325fd7259c41b59541393dc/vscode-sbt-scala/client/package.json#L21. Also make sure you don't have another language server running at the same time like vscode-dotty or https://github.com/dragos/dragos-vscode-scala, they won't work together as I explained in: https://github.com/sbt/sbt/pull/3524#discussion_r139563361

**Comment by eed3si9n** [comment]
ok. I am going to merge this and release a milestone.

**Comment by kiritsuku** [comment]
I checked it all but except of the initialization and shutdown routine nothing works. I need to check if I can get the dotty extension up and running, maybe I'm still using something incorrectly.





---- On Tue, 03 Oct 2017 17:48:20 +0200 Guillaume Martres &lt;notifications@github.com&gt; wrote ----




@sschaef: Right now it's restricted to .scala and .java files: https://github.com/eed3si9n/sbt/blob/1ad9360e3e9446eed325fd7259c41b59541393dc/vscode-sbt-scala/client/src/extension.ts and requires a build.sbt file: https://github.com/eed3si9n/sbt/blob/1ad9360e3e9446eed325fd7259c41b59541393dc/vscode-sbt-scala/client/package.json#L21. Also make sure you don't have another language server running at the same time like vscode-dotty or https://github.com/dragos/dragos-vscode-scala, they won't work together as I explained in: #3524 (comment)

**Comment by wpopielarski** [comment]
@sschaef "Ok, I fixed the problem. On sbt startup there is a npm install run on the vscode project. I removed the node_modules folder and restart sbt. Now everything works perfectly fine."
the `target` folder in `vscode-sbt-scala` needs to removed too

**Comment by ghisvail** [comment]
Any chance to get something similar for Atom?

Version 1.21 got support for the language server protocol and is already used by a number of IDE integration packages such as `ide-typescript` or `ide-java`. I can picture sbt 1.1.x providing similar functionalities in a future `ide-scala` package.

**Comment by wpopielarski** [comment]
Well, I'm wonder if in Atom you can load vscode extensions :). If not Atom
specific is needed I think with static syntax highlighting in
`Scala.tmLanguage`

2017-10-06 10:19 GMT+02:00 Ghislain Antony Vaillant <
notifications@github.com>:

**Pull request merged**